src/sentry/utils/sentryappwebhookrequests.py

```diff
@@ -0,0 +1,128 @@
+from __future__ import absolute_import
+
+import six
+import json
+
+from dateutil.parser import parse as parse_date
+from django.conf import settings
+from django.utils import timezone
+
+from sentry.utils import redis
+from sentry.models.sentryapp import VALID_EVENTS
+
+
+BUFFER_SIZE = 100
+
+
+class SentryAppWebhookRequestsBuffer(object):
```

Comment:
What is cleaning up this buffer? I understand that per sentry_app, it's capped at 100. But there's no expiry on the keys or anything to prevent the set of buffers from growing forever that I can see. We're just infinitely going to be creating new lists.

We need to leverage EXPIRE as well: https://redis.io/commands/expire

This way we don't forever keep data.

And also potentially, not required, explicit deletion of the keys in here when a SentryApp is deleted.

Reply:
@mattrobenolt For each sentry_app, there won't be infinite amount of lists because we are creating only twice as many lists as there are webhook event types, which is finite (right now there's 5 but has the potential to be more but not infinitely more). The only unbounded thing is the number of sentry_apps but that will ultimately just be however many Sentry Apps exist out there

is it unfeasible to keep the data "forever" in that it doesn't expire? I wanted to show the last 100 errors/requests regardless of how long ago they were, but maybe if they are older than a certain time that's okay too

Reply:
Sure, but they are inifinite in that there's no cap on the amount of sentry_apps there might be. I understand there are a very small number today.

More the concern is when an app is deleted, this will be left behind. So maybe if we just focus on that for now to make sure if an app goes away, the redis buffer also goes away, taht'd be a good first step.

But I'd also argue that an expiry of like, 30 days wouldn't be a bad experience. If there's no error in 30 days... I feel like the data isn't that relevant anymore and wouldn't lead to a bad experience.

Going a bit more off the extreme end here, it might be a little bit more effective to do this more similar to tsdb style where there's a buffer per day (or some interval), and each buffer is ttl'd per day. Then over 30 days, there's a capped list of 100 per day, but each day gets expired. Or something, just throwing out ideas or ways this can be done.

Either way, I think it's important that we hook into some sort of deletion to make sure these keys go away. Either through explicit deletions and/or expirations. Relying on deletions is tricky since they are two separate datastores, and there's nothing to guarantee they are left in sync. So if a database row were deleted, but the redis key wasn't, the redis key lives forever. The only way we'd find it again is to iterate all keys in redis and find it, which is why expiration is usually a good compliment to make sure if a deletion fails, it eventually expires away.

Reply:
@mattrobenolt What are the two separate datastores in this case?

And yeah I'm thinking a 30 day expiry date is okay for all these keys :)

Reply:
> What are the two separate datastores in this case?

Redis and Postgres. There's no foreign keys or anything to strongly guarantee consistency between these datastores. So if you delete from Postgres the row, the key in Redis might not get deleted. Leaving it in Redis forever. The only option to attempt to keep these in sync is through a Django signal, but that's application level. Imagine Postgres being online, but Redis server is down. What happens? The row gets deleted from Postgres, but it errored when deleting from Redis. Or, it's possible to lose operations against Redis in a failure scenario. Redis might accept the DELETE on the key, but then if the process crashes, the DELETE might not have persisted to the AOF or the RDB dump yet, meaning the key will come back when it restarts.

These aren't highly probable scenarios, but they are plausible scenarios and we generally like to account for them in some way.

Reply:
> What are the two separate datastores in this case?

Redis and Postgres. There's no foreign keys or anything to strongly guarantee consistency between these datastores. So if you delete from Postgres the row, the key in Redis might not get deleted. Leaving it in Redis forever. The only option to attempt to keep these in sync is through a Django signal, but that's application level. Imagine Postgres being online, but Redis server is down. What happens? The row gets deleted from Postgres, but it errored when deleting from Redis. Or, it's possible to lose operations against Redis in a failure scenario. Redis might accept the DELETE on the key, but then if the process crashes, the DELETE might not have persisted to the AOF or the RDB dump yet, meaning the key will come back when it restarts.

These aren't highly probable scenarios, but they are plausible scenarios and we generally like to account for them in some way.

Reply:
@mattrobenolt Wait the information here is just stored in Redis, not Postgres unless I'm missing something here. It used to be stored in Postgres but this PR is supposed to introduce storing these requests/errors in Redis in order to move away from Postgres for this

Reply:
Right, but SentryApp is in Postgres. Which is effectively what this is linking to. So if the app is deleted, this data persists in Redis.

Reply:
Ah okay I understand now :) I'll implement the 30 day expiry period

Reply:
*sorry actually after talking with Steve we decided on 90 days

Reply:
I wanna make sure we're on the same page here in terms of how this would work, because 90 days seems a bit excessive to me.

Right now, when a new event happens, you're doing the following:

```
LPUSH key value
LTRIM key 100
```

I'm suggesting:

```
LPUSH key value
LTRIM key 100
EXPIRE {30 days}
```

So what'll happen here is that every time a new item is pushed in, we reset the expiry to 30 days from that point. Which means they key will only expire if it never sees another error within 30 days. If data keeps coming in regularly it'll stay forever, until data stops flowing in.

I'd strongly argue, if there's an error within an app, and it's over 30 days over without ever happening again, it can't be useful to anyone. I'd even argue that over 10 days is probably useless data, but 30 is reasonable.